{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute Interaction Distances between variants and Ligands, PTMs, Drugs\n",
    "Creates a dataframe of all available interaction data using PDB, MMTF-pySpark, UniProt, and databases like dbPTM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<style>\n",
       "    div#notebook-container    { width: 95%; }\n",
       "    div#menubar-container     { width: 65%; }\n",
       "    div#maintoolbar-container { width: 99%; }\n",
       "</style>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import py3Dmol\n",
    "from ipywidgets import interact, IntSlider\n",
    "import pandas as pd\n",
    "pd.set_option('display.expand_frame_repr', False)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import interactive\n",
    "interactive(True)\n",
    "import seaborn as sns\n",
    "sns.set(style=\"whitegrid\")\n",
    "\n",
    "import ast\n",
    "from sklearn.metrics.pairwise import euclidean_distances\n",
    "\n",
    "from IPython.display import display, HTML\n",
    "\n",
    "display(HTML(data=\"\"\"\n",
    "<style>\n",
    "    div#notebook-container    { width: 95%; }\n",
    "    div#menubar-container     { width: 65%; }\n",
    "    div#maintoolbar-container { width: 99%; }\n",
    "</style>\n",
    "\"\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark.sql import SparkSession\n",
    "from mmtfPyspark.io import mmtfReader\n",
    "from mmtfPyspark.interactions import InteractionExtractor, InteractionFilter\n",
    "\n",
    "from mmtfPyspark.datasets import pdbToUniProt\n",
    "from mmtfPyspark.utils import traverseStructureHierarchy, ColumnarStructure\n",
    "from mmtfPyspark.datasets import drugBankDataset, pdbjMineDataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run /home/ebc/variome/GIT/mmtf-pyspark/mmtfPyspark/interactions/ExcludedLigandSets.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize Spark\n",
    "spark = SparkSession.builder.master(\"local[4]\").appName(\"2C-MapLigandInteractions\").getOrCreate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### input parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file_name = '../analysis/NRF2_pathway/dataframes/step2/mutations_NRF2v2_step2B.csv' # mutations and PDB structures\n",
    "\n",
    "output_file_name = '../analysis/NRF2_pathway/dataframes/step2/mutations_NRF2v2_step2C.csv' # data on interaction distances\n",
    "\n",
    "path_plots = '../analysis/NRF2_pathway/plots/step2/'\n",
    "path_df = '../analysis/NRF2_pathway/dataframes/step2/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Protein_Change</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>variationId</th>\n",
       "      <th>uniprotId</th>\n",
       "      <th>structureChainId</th>\n",
       "      <th>targetGroupId</th>\n",
       "      <th>targetChainId</th>\n",
       "      <th>targetGroupNumber</th>\n",
       "      <th>cluster_x</th>\n",
       "      <th>cluster_y</th>\n",
       "      <th>cluster_z</th>\n",
       "      <th>atom_nom</th>\n",
       "      <th>PDB_CHAIN_ID</th>\n",
       "      <th>SP_PRIMARY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6447</td>\n",
       "      <td>UCHL1</td>\n",
       "      <td>p.P43S</td>\n",
       "      <td>HCT15_LARGE_INTESTINE</td>\n",
       "      <td>chr4:g.41259707C&gt;T</td>\n",
       "      <td>P09936</td>\n",
       "      <td>3KW5.A</td>\n",
       "      <td>PRO</td>\n",
       "      <td>A</td>\n",
       "      <td>43</td>\n",
       "      <td>36.817001</td>\n",
       "      <td>56.400002</td>\n",
       "      <td>95.001999</td>\n",
       "      <td>CG</td>\n",
       "      <td>3KW5.A</td>\n",
       "      <td>P09936</td>\n",
       "      <td>UCHL1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6448</td>\n",
       "      <td>UCHL1</td>\n",
       "      <td>p.V200V</td>\n",
       "      <td>LP1_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>chr4:g.41270018C&gt;A</td>\n",
       "      <td>P09936</td>\n",
       "      <td>3KW5.A</td>\n",
       "      <td>VAL</td>\n",
       "      <td>A</td>\n",
       "      <td>200</td>\n",
       "      <td>43.471001</td>\n",
       "      <td>63.093002</td>\n",
       "      <td>75.249001</td>\n",
       "      <td>CG2_CG1</td>\n",
       "      <td>3KW5.A</td>\n",
       "      <td>P09936</td>\n",
       "      <td>UCHL1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index Hugo_Symbol Protein_Change                    Tumor_Sample_Barcode         variationId uniprotId structureChainId targetGroupId targetChainId  targetGroupNumber  cluster_x  cluster_y  cluster_z atom_nom PDB_CHAIN_ID SP_PRIMARY   GENE  coverage\n",
       "0   6447       UCHL1         p.P43S                   HCT15_LARGE_INTESTINE  chr4:g.41259707C>T    P09936           3KW5.A           PRO             A                 43  36.817001  56.400002  95.001999       CG       3KW5.A     P09936  UCHL1       1.0\n",
       "1   6448       UCHL1        p.V200V  LP1_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE  chr4:g.41270018C>A    P09936           3KW5.A           VAL             A                200  43.471001  63.093002  75.249001  CG2_CG1       3KW5.A     P09936  UCHL1       1.0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(input_file_name,index_col=0)\n",
    "df.reset_index(inplace=True)\n",
    "df.head(2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute distance of variants to nearest PTM site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of PTMs for gene set: 326\n"
     ]
    }
   ],
   "source": [
    "# read in PTM data\n",
    "ptm_data = pd.read_table('../data/dbPTM/Phosphorylation.txt',header=None)\n",
    "ptm_data = ptm_data[ptm_data[1].isin(df.uniprotId.unique().tolist())]\n",
    "ptm_data.columns = ['gene','uniprotId','residue','type','?','seq']\n",
    "\n",
    "\n",
    "ubq_data = pd.read_table('../data/dbPTM/Ubiquitination.txt',header=None)\n",
    "ubq_data = ubq_data[ubq_data[1].isin(df.uniprotId.unique().tolist())]\n",
    "ubq_data.columns = ['gene','uniprotId','residue','type','?','seq']\n",
    "\n",
    "ptm_data = pd.concat([ptm_data,ubq_data])\n",
    "print(\"Total number of PTMs for gene set:\",len(ptm_data))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get X Y Z coordinates of PTM residue per PDB "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapping of UniProt to PDB\n",
    "up = pdbToUniProt.get_cached_residue_mappings().filter(\"pdbResNum IS NOT NULL\").filter(\"uniprotNum IS NOT NULL\")\n",
    "up_map = up.filter(up.uniprotId.isin(ptm_data.uniprotId.tolist())).toPandas()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Steps:\n",
    "# 1) Select PTM residues that have PDB data available\n",
    "# 2) Merge PTM info with UniProt and PDB info\n",
    "\n",
    "ptm_residues = pd.DataFrame()\n",
    "\n",
    "for u in ptm_data.uniprotId.unique():\n",
    "    df1 = ptm_data[ptm_data.uniprotId==u]\n",
    "    df2 = up_map[up_map.uniprotId==u]\n",
    "\n",
    "    for i in df1.index:\n",
    "        ures = df1.loc[i,'residue']\n",
    "        change = df1.loc[i,'type']\n",
    "        seq = df1.loc[i,'seq']\n",
    "\n",
    "        tmp = df2[df2.uniprotNum==ures]\n",
    "        tmp.reset_index(inplace=True)\n",
    "        \n",
    "        if len(tmp)>0:\n",
    "            for j in tmp.index:\n",
    "                if j==0:\n",
    "                    tmp.insert(0, 'PtmType', change)\n",
    "                    tmp.insert(0, 'seq', seq)\n",
    "                else:\n",
    "                    tmp.at[j,'PtmType'] = change\n",
    "                    tmp.at[j,'seq'] = seq\n",
    "            ptm_residues = pd.concat([ptm_residues, tmp])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# 3) Routine to retrieve the XYZ coords and other information from PDB\n",
    "\n",
    "def get_PTM_pdb_data(u,p):\n",
    "    \n",
    "    if '.' in p:\n",
    "        p = p.split('.')[0]\n",
    "    \n",
    "    pdb_data = []\n",
    "\n",
    "    #gets PDB file and columnar structure attributes\n",
    "    pdb = mmtfReader.download_full_mmtf_files([p])\n",
    "    structure = pdb.values().first()\n",
    "    arrays = ColumnarStructure(structure, firstModelOnly=True)\n",
    "    x = arrays.get_x_coords()\n",
    "    y = arrays.get_y_coords()\n",
    "    z = arrays.get_z_coords()\n",
    "    group_names = arrays.get_group_names()\n",
    "    group_numbers = arrays.get_group_numbers()\n",
    "    chain_names = arrays.get_chain_names()\n",
    "    group_indices = arrays.get_group_to_atom_indices()\n",
    "\n",
    "    for i in range(arrays.get_num_groups()):   \n",
    "        start = group_indices[i]\n",
    "        end = group_indices[i+1]\n",
    "\n",
    "        # match residue number in PDB with PTM residue\n",
    "        if group_numbers[start] in ptm_residues[ptm_residues.structureChainId.str.startswith(p)].pdbResNum.unique().tolist():\n",
    "        #if (int(group_numbers[start])>120) and (int(group_numbers[start])<129):\n",
    "            #print(group_names[start] + \"-\" + chain_names[start] + \":\" + group_numbers[start] + \" \" + str(start) + \" - \" + str(end) + \"   \" +str(x[start:end]))\n",
    "            pdb_data.append({'uniprotId':u,'pdbResName':group_names[start],'structureChainId':p+'.'+chain_names[start],'pdbResNum':str(group_numbers[start]),'atom_idx_start':int(start),'atom_idx_end':int(end),'x_list':x[start:end],'y_list':y[start:end],'z_list':z[start:end]})\n",
    "\n",
    "            \n",
    "    tmp = pd.DataFrame(pdb_data)  \n",
    "    ptm_pdb_data = pd.merge(tmp,ptm_residues[ptm_residues.structureChainId.str.startswith(p)], on=['uniprotId','pdbResNum','structureChainId'])\n",
    "    \n",
    "    return (ptm_pdb_data)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "# Routine to retrieve the XYZ coords and other information from PDB -- new, need to update mmtf-pyspark\n",
    "\n",
    "def get_PTM_pdb_data(u,p):\n",
    "    \n",
    "    pdb_data = pd.DataFrame()\n",
    "    \n",
    "    if '.' in p:\n",
    "        pp = p.split('.')[0]\n",
    "    else:\n",
    "        pp = p\n",
    "    \n",
    "    structures = mmtfReader.download_mmtf_files([pp]).cache()\n",
    "\n",
    "    # get first structure from Spark RDD (keys = PDB IDs, value = mmtf structures)\n",
    "    first_structure = structures.values().first()\n",
    "\n",
    "    # convert to a Pandas dataframe\n",
    "    df = ColumnarStructure(first_structure).to_pandas()\n",
    "    \n",
    "    for res in ptm_residues[ptm_residues.structureChainId.str.startswith(pp)].pdbResNum.unique():\n",
    "\n",
    "        query = \"polymer and (group_number = %s )\"%res\n",
    "        df_tmp = df.query(query)\n",
    "        df_tmp['uniprotId'] = [u for i in df_tmp.index]\n",
    "        df_tmp['structureChainId'] = [p for i in df_tmp.index]\n",
    "        df_tmp.rename({'group_number':'pdbResNum'})\n",
    "        pdb_data = pd.concat([pdb_data,df_tmp])\n",
    "    \n",
    "        #pdb_data.append({'uniprotId':u,'pdbResName':group_names[start],'structureChainId':p+'.'+chain_names[start],'pdbResNum':str(group_numbers[start]),'atom_idx_start':int(start),'atom_idx_end':int(end),'x_list':x[start:end],'y_list':y[start:end],'z_list':z[start:end]})\n",
    "\n",
    "    ptm_pdb_data = pd.merge(pd.DataFrame(pdb_data),ptm_residues[ptm_residues.structureChainId.str.startswith(p)], on=['uniprotId','pdbResNum','structureChainId'])\n",
    "    return (ptm_pdb_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level_0</th>\n",
       "      <th>atom_idx_end</th>\n",
       "      <th>atom_idx_start</th>\n",
       "      <th>pdbResName</th>\n",
       "      <th>pdbResNum</th>\n",
       "      <th>structureChainId</th>\n",
       "      <th>uniprotId</th>\n",
       "      <th>x_list</th>\n",
       "      <th>y_list</th>\n",
       "      <th>z_list</th>\n",
       "      <th>seq</th>\n",
       "      <th>PtmType</th>\n",
       "      <th>index</th>\n",
       "      <th>pdbSeqNum</th>\n",
       "      <th>uniprotNum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>203</td>\n",
       "      <td>197</td>\n",
       "      <td>SER</td>\n",
       "      <td>23</td>\n",
       "      <td>4ICC.X</td>\n",
       "      <td>O60218</td>\n",
       "      <td>[-1.774, -1.918, -0.636, -0.295, -3.129, -4.306]</td>\n",
       "      <td>[35.316, 36.393, 36.513, 35.602, 36.143, 36.179]</td>\n",
       "      <td>[2.715, 3.699, 4.548, 5.303, 4.586, 3.79]</td>\n",
       "      <td>MPIVGLGTWKSPLGKVKEAVK</td>\n",
       "      <td>Phosphorylation</td>\n",
       "      <td>7640</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>319</td>\n",
       "      <td>TYR</td>\n",
       "      <td>40</td>\n",
       "      <td>4ICC.X</td>\n",
       "      <td>O60218</td>\n",
       "      <td>[-17.501, -17.076, -16.955, -16.439, -15.693, ...</td>\n",
       "      <td>[37.745, 36.458, 35.448, 35.772, 36.545, 37.25...</td>\n",
       "      <td>[10.468, 9.909, 11.044, 12.119, 9.275, 7.948, ...</td>\n",
       "      <td>EAVKVAIDAGYRHIDCAYVYQ</td>\n",
       "      <td>Phosphorylation</td>\n",
       "      <td>7657</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   level_0  atom_idx_end  atom_idx_start pdbResName pdbResNum structureChainId uniprotId                                             x_list                                             y_list                                             z_list                    seq          PtmType  index  pdbSeqNum  uniprotNum\n",
       "0        0           203             197        SER        23           4ICC.X    O60218   [-1.774, -1.918, -0.636, -0.295, -3.129, -4.306]   [35.316, 36.393, 36.513, 35.602, 36.143, 36.179]          [2.715, 3.699, 4.548, 5.303, 4.586, 3.79]  MPIVGLGTWKSPLGKVKEAVK  Phosphorylation   7640         23          23\n",
       "1        1           331             319        TYR        40           4ICC.X    O60218  [-17.501, -17.076, -16.955, -16.439, -15.693, ...  [37.745, 36.458, 35.448, 35.772, 36.545, 37.25...  [10.468, 9.909, 11.044, 12.119, 9.275, 7.948, ...  EAVKVAIDAGYRHIDCAYVYQ  Phosphorylation   7657         40          40"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_PTM_PDB = pd.DataFrame()\n",
    "\n",
    "for u in ptm_residues.uniprotId.unique():\n",
    "    for p in ptm_residues[ptm_residues.uniprotId==u].structureChainId.unique():\n",
    "        \n",
    "        tmp = get_PTM_pdb_data(u,p)\n",
    "        DF_PTM_PDB = pd.concat([DF_PTM_PDB,tmp])\n",
    "        \n",
    "DF_PTM_PDB.reset_index(inplace=True)\n",
    "DF_PTM_PDB.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_PTM_PDB.to_csv(path_df+'PTM_PDB_data.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compute nearest-neighbor distances from variants to nearby PTM sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No PTM data for gene Q16620\n",
      "No PTM data for gene Q16620\n",
      "No PTM data for gene Q16620\n",
      "No PTM data for gene Q16620\n",
      "No PTM data for gene Q16620\n",
      "No PTM data for gene Q16620\n",
      "No PTM data for gene Q16620\n",
      "No PTM data for gene Q16620\n",
      "No PTM data for gene Q16620\n",
      "No PTM data for gene Q16620\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n",
      "No PTM data for gene Q16236\n"
     ]
    }
   ],
   "source": [
    "## Note: may need to consider biological assemblies - use mmtf interaction tools\n",
    "\n",
    "ptm_res = []\n",
    "ptm_group = []\n",
    "ptm_desc = []\n",
    "ptm_dist = []\n",
    "\n",
    "for i in df.index:\n",
    "    \n",
    "    var_x = df.loc[i,'cluster_x']\n",
    "    var_y = df.loc[i,'cluster_y']\n",
    "    var_z = df.loc[i,'cluster_z']\n",
    "    X = [var_x,var_y,var_z]\n",
    "    \n",
    "    p = df.loc[i,'structureChainId']\n",
    "    u = df.loc[i,'uniprotId']\n",
    "    \n",
    "    tmp_ptm_df = DF_PTM_PDB[(DF_PTM_PDB['uniprotId']==u)&(DF_PTM_PDB['structureChainId']==p)]\n",
    "    \n",
    "    # loop over PTMs for that specific PDB/protein\n",
    "    \n",
    "    PTM_dict = {}\n",
    "    nearest_neighbor = ''\n",
    "    \n",
    "    # finds nearest PTM to variant\n",
    "    for j in tmp_ptm_df.index:\n",
    "        ptm_x = tmp_ptm_df.loc[j,'x_list']\n",
    "        ptm_y = tmp_ptm_df.loc[j,'y_list']\n",
    "        ptm_z = tmp_ptm_df.loc[j,'z_list']\n",
    "        \n",
    "        ptm_resname = tmp_ptm_df.loc[j,'pdbResName']\n",
    "        ptm_num = tmp_ptm_df.loc[j,'pdbResNum']\n",
    "        ptm_type = tmp_ptm_df.loc[j,'PtmType']\n",
    "        \n",
    "        Y = []\n",
    "        \n",
    "        for z in range(0,len(ptm_x)):\n",
    "            # recombine lists\n",
    "            Y.append([ptm_x[z],ptm_y[z],ptm_z[z]])\n",
    "        \n",
    "        # compute pairwise euclidean distance and take the minimum distance\n",
    "        dist = euclidean_distances([X], Y)\n",
    "        min_dist = np.min(dist)  # select distance of closest atom on the PTM to variant x,y,z\n",
    "        \n",
    "        PTM_dict[j] = (ptm_resname, ptm_num, ptm_type, min_dist)\n",
    "        \n",
    "        if nearest_neighbor == '':\n",
    "            nearest_neighbor = PTM_dict[j]\n",
    "        else:\n",
    "            if PTM_dict[j][3] < nearest_neighbor[3]:\n",
    "                nearest_neighbor = PTM_dict[j]\n",
    "                \n",
    "    try:\n",
    "        ptm_res.append(nearest_neighbor[0])\n",
    "        ptm_group.append(nearest_neighbor[1])\n",
    "        ptm_desc.append(nearest_neighbor[2])\n",
    "        ptm_dist.append(nearest_neighbor[3])    \n",
    "    except IndexError:\n",
    "        print(\"No PTM data for gene %s\"%u)\n",
    "        ptm_res.append('')\n",
    "        ptm_group.append('')\n",
    "        ptm_desc.append('')\n",
    "        ptm_dist.append(1000)   \n",
    "        \n",
    "        \n",
    "df['PTM_dist'] = ptm_dist\n",
    "df['PTM_type'] = ptm_desc\n",
    "df['PTM_resName'] = ptm_res\n",
    "df['PTM_resNum'] = ptm_group\n",
    "           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Protein_Change</th>\n",
       "      <th>PTM_dist</th>\n",
       "      <th>PTM_resNum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UCHL1</td>\n",
       "      <td>SNU1040_LARGE_INTESTINE</td>\n",
       "      <td>p.K83K</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UCHL1</td>\n",
       "      <td>MCC26_SKIN</td>\n",
       "      <td>p.S89F</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>VMCUB1_URINARY_TRACT</td>\n",
       "      <td>p.P28A</td>\n",
       "      <td>3.232057</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>KYAE1_OESOPHAGUS</td>\n",
       "      <td>p.V2V</td>\n",
       "      <td>4.710373</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>MDAMB361_BREAST</td>\n",
       "      <td>p.E90Q</td>\n",
       "      <td>4.430407</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>S117_SOFT_TISSUE</td>\n",
       "      <td>p.K112K</td>\n",
       "      <td>4.165546</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>CAL33_UPPER_AERODIGESTIVE_TRACT</td>\n",
       "      <td>p.K57K</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>NCIH684_LIVER</td>\n",
       "      <td>p.I168T</td>\n",
       "      <td>4.105054</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>DV90_LUNG</td>\n",
       "      <td>p.A116T</td>\n",
       "      <td>2.940409</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>T173_FIBROBLAST</td>\n",
       "      <td>p.L229L</td>\n",
       "      <td>4.766182</td>\n",
       "      <td>284</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Hugo_Symbol             Tumor_Sample_Barcode Protein_Change  PTM_dist PTM_resNum\n",
       "3        UCHL1          SNU1040_LARGE_INTESTINE         p.K83K  0.000000         83\n",
       "4        UCHL1                       MCC26_SKIN         p.S89F  0.000000         89\n",
       "11      TXNRD1             VMCUB1_URINARY_TRACT         p.P28A  3.232057         67\n",
       "13      TXNRD1                 KYAE1_OESOPHAGUS          p.V2V  4.710373        127\n",
       "20      TXNRD1                  MDAMB361_BREAST         p.E90Q  4.430407        152\n",
       "29      TXNRD1                 S117_SOFT_TISSUE        p.K112K  4.165546        152\n",
       "30      TXNRD1  CAL33_UPPER_AERODIGESTIVE_TRACT         p.K57K  0.000000         95\n",
       "31      TXNRD1                    NCIH684_LIVER        p.I168T  4.105054        157\n",
       "33      TXNRD1                        DV90_LUNG        p.A116T  2.940409         13\n",
       "41      TXNRD1                  T173_FIBROBLAST        p.L229L  4.766182        284"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.PTM_dist<5][['Hugo_Symbol','Tumor_Sample_Barcode','Protein_Change','PTM_dist','PTM_resNum']].drop_duplicates().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute Distance of Variants to Ligand Binding Site"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "note: will have to consider PDB-PDB mapping or selection of PDB files with ligands before hand in step 3\n",
    "\n",
    "note -2: would be nice to change to consider interactions with multiple ligands"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdb_ids = [i.split('.')[0] for i in df.structureChainId.unique()]\n",
    "structures = mmtfReader.download_mmtf_files(pdb_ids)\n",
    "structures = structures.filter(lambda s: s[1].num_models == 1)  ## Currently, only structures with 1 model are supported"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "distance_cutoff = 17\n",
    "\n",
    "exclude = [\"HOH\", \"DOD\",\"ACT\"] + list(JUNK)\n",
    "\n",
    "interaction_filter = InteractionFilter(distanceCutoff=distance_cutoff)\n",
    "interaction_filter.set_query_groups(False, exclude)  # exclude junk and water\n",
    "\n",
    "Lig_interactions = InteractionExtractor().get_ligand_polymer_interactions(structures, interaction_filter, level='atom').toPandas()\n",
    "\n",
    "Lig_interactions['targetGroupNumber'] = Lig_interactions['targetGroupNumber'].astype(int)\n",
    "\n",
    "# Lig_interactions.to_csv(path_df+'ligand_info.csv') # caution large file size > 100mb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "lig_resName = []\n",
    "lig_dist = []\n",
    "lig_resNum = []\n",
    "\n",
    "for i in df.index:\n",
    "    \n",
    "    p = df.loc[i,'structureChainId']\n",
    "    t = df.loc[i,'targetGroupNumber']\n",
    "    \n",
    "    r = ''\n",
    "    d = 1000\n",
    "    n = ''\n",
    "    \n",
    "    tmp_lig_df = Lig_interactions[(Lig_interactions['structureChainId']==p)&(Lig_interactions.targetGroupNumber==t)].sort_values('distance',ascending=True)\n",
    "    tmp_lig_df.reset_index(inplace=True)\n",
    "    \n",
    "    # TO DO: opt for additional ligands, not just nearest neighboring one\n",
    "    #if len(tmp_lig_df) > 0:\n",
    "    #    for j in tmp_lig_df.queryGroupId.unique():\n",
    "    #        tmp = tmp_lig_df[tmp_lig_df.queryGroupId==j].sort_values('distance',ascending=True)\n",
    "        \n",
    "    if len(tmp_lig_df) > 0:\n",
    "        \n",
    "        r = tmp_lig_df.loc[0,'queryGroupId']\n",
    "        d = tmp_lig_df.loc[0,'distance']\n",
    "        n = tmp_lig_df.loc[0,'queryGroupNumber']\n",
    "                   \n",
    "        try:\n",
    "            lig_resName.append(r)\n",
    "            lig_dist.append(d)\n",
    "            lig_resNum.append(n)\n",
    "        except IndexError:\n",
    "            lig_resName.append('')\n",
    "            lig_dist.append(1000)\n",
    "            lig_resNum.append('')\n",
    "            \n",
    "    else:\n",
    "        lig_resName.append('')\n",
    "        lig_dist.append(1000)\n",
    "        lig_resNum.append('')\n",
    "        \n",
    "        \n",
    "df['Ligand_dist'] = lig_dist\n",
    "df['Ligand_resName'] = lig_resName\n",
    "df['Ligand_resNum'] = lig_resNum\n",
    "           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Protein_Change</th>\n",
       "      <th>Ligand_dist</th>\n",
       "      <th>Ligand_resName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UCHL1</td>\n",
       "      <td>MCC26_SKIN</td>\n",
       "      <td>p.S89F</td>\n",
       "      <td>4.458467</td>\n",
       "      <td>GVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>GRST_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>p.T335T</td>\n",
       "      <td>4.987658</td>\n",
       "      <td>FAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>MOLT16_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>p.T305T</td>\n",
       "      <td>3.126904</td>\n",
       "      <td>FAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>NCIH684_LIVER</td>\n",
       "      <td>p.I168T</td>\n",
       "      <td>3.739374</td>\n",
       "      <td>FAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>HUH1_LIVER</td>\n",
       "      <td>p.A122V</td>\n",
       "      <td>3.230645</td>\n",
       "      <td>FAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>HEC1_ENDOMETRIUM</td>\n",
       "      <td>p.H434R</td>\n",
       "      <td>2.573166</td>\n",
       "      <td>GOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>AGS_STOMACH</td>\n",
       "      <td>p.N331D</td>\n",
       "      <td>2.349676</td>\n",
       "      <td>GOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>SNU1040_LARGE_INTESTINE</td>\n",
       "      <td>p.Y93Y</td>\n",
       "      <td>3.183022</td>\n",
       "      <td>FAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>SRXN1</td>\n",
       "      <td>HCC366_LUNG</td>\n",
       "      <td>p.R101H</td>\n",
       "      <td>2.530788</td>\n",
       "      <td>PO4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>SRXN1</td>\n",
       "      <td>SNU349_KIDNEY</td>\n",
       "      <td>p.Y102Y</td>\n",
       "      <td>4.789872</td>\n",
       "      <td>PO4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>SIRT1</td>\n",
       "      <td>BICR18_UPPER_AERODIGESTIVE_TRACT</td>\n",
       "      <td>p.C395C</td>\n",
       "      <td>2.257232</td>\n",
       "      <td>ZN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>RAB6B</td>\n",
       "      <td>NCIH630_LARGE_INTESTINE</td>\n",
       "      <td>p.D39N</td>\n",
       "      <td>3.055144</td>\n",
       "      <td>GSP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Hugo_Symbol                       Tumor_Sample_Barcode Protein_Change  Ligand_dist Ligand_resName\n",
       "4         UCHL1                                 MCC26_SKIN         p.S89F     4.458467            GVE\n",
       "27       TXNRD1    GRST_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE        p.T335T     4.987658            FAD\n",
       "28       TXNRD1  MOLT16_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE        p.T305T     3.126904            FAD\n",
       "31       TXNRD1                              NCIH684_LIVER        p.I168T     3.739374            FAD\n",
       "34       TXNRD1                                 HUH1_LIVER        p.A122V     3.230645            FAD\n",
       "52       TXNRD1                           HEC1_ENDOMETRIUM        p.H434R     2.573166            GOL\n",
       "54       TXNRD1                                AGS_STOMACH        p.N331D     2.349676            GOL\n",
       "59       TXNRD1                    SNU1040_LARGE_INTESTINE         p.Y93Y     3.183022            FAD\n",
       "60        SRXN1                                HCC366_LUNG        p.R101H     2.530788            PO4\n",
       "67        SRXN1                              SNU349_KIDNEY        p.Y102Y     4.789872            PO4\n",
       "68        SIRT1           BICR18_UPPER_AERODIGESTIVE_TRACT        p.C395C     2.257232             ZN\n",
       "103       RAB6B                    NCIH630_LARGE_INTESTINE         p.D39N     3.055144            GSP"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.Ligand_dist<5][['Hugo_Symbol','Tumor_Sample_Barcode','Protein_Change','Ligand_dist','Ligand_resName']].drop_duplicates().head(12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute Distances of Variants to Bound Drugs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get InChiKey for ligands in PDB with molecular weight >= 250"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ligand_id</th>\n",
       "      <th>formula_weight</th>\n",
       "      <th>inchi_key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PQQ</td>\n",
       "      <td>330.206</td>\n",
       "      <td>MMXZSJMASHPLLR-UHFFFAOYSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RKS</td>\n",
       "      <td>540.563</td>\n",
       "      <td>WXZHDJXCVBLPHY-MOPGFXCFSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GR6</td>\n",
       "      <td>630.749</td>\n",
       "      <td>ZQJBDIZBNGQJDT-JZAADHNPSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B83</td>\n",
       "      <td>444.522</td>\n",
       "      <td>YMZMFFRGSWCLPY-KDURUIRLSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3P1</td>\n",
       "      <td>423.425</td>\n",
       "      <td>NAHSCHKAPXMNFP-NVQRDWNXSA-N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ligand_id  formula_weight                    inchi_key\n",
       "0       PQQ         330.206  MMXZSJMASHPLLR-UHFFFAOYSA-N\n",
       "1       RKS         540.563  WXZHDJXCVBLPHY-MOPGFXCFSA-N\n",
       "2       GR6         630.749  ZQJBDIZBNGQJDT-JZAADHNPSA-N\n",
       "3       B83         444.522  YMZMFFRGSWCLPY-KDURUIRLSA-N\n",
       "4       3P1         423.425  NAHSCHKAPXMNFP-NVQRDWNXSA-N"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mw_min = 250  # minimum molecular weight for drug molecules\n",
    "\n",
    "ccQuery = \"SELECT c.id as ligand_id, c.formula_weight, d.descriptor as inchi_key FROM pdbj.chem_comp c \\\n",
    "           JOIN cc.pdbx_chem_comp_descriptor d ON d.comp_id = c.id \\\n",
    "           WHERE d.type = 'InChIKey' AND c.formula_weight >= \" + str(mw_min)\n",
    "\n",
    "ligands = pdbjMineDataset.get_dataset(ccQuery).dropDuplicates().toPandas()\n",
    "ligands.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Join dataset on ligand id to add InchiKeys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt = Lig_interactions.merge(ligands, left_on=['queryGroupId'], right_on=['ligand_id'], how='inner')\n",
    "mt = mt.drop_duplicates()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download open DrugBank dataset¶"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We use the open DrugBank dataset. One disadvantage of the open DrugBank dataset is that it not only contains approved drugs, but many other compounds in pharmaceutical use such as ethanol, ATP, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "drugs = drugBankDataset.get_open_drug_links() \\\n",
    "                       .filter(\"StandardInChIKey IS NOT NULL\") \\\n",
    "                       .filter(\"CAS IS NOT NULL\") \\\n",
    "                       .toPandas()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The DrugBank password protected datasets contain more information (e.g., approval status). To use these datasets, you need to create a free DrugBank account and supply username/passwork to access these datasets. \n",
    "\n",
    "For this demo, we continue with the open drug bank dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "username = \"ebrunk@ucsd.edu\"\n",
    "password = \"ssbio123\"\n",
    "drugs = drugBankDataset.get_drug_links(\"APPROVED\", username,password) \\\n",
    "                        .filter(\"InChIKey IS NOT NULL\") \\\n",
    "                        .toPandas()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "username = \"<your DrugBank account username>\"\n",
    "password = \"<your DrugBank account password>\"\n",
    "drugs = drugBankDataset.get_drug_links(\"APPROVED\", username,password) \\\n",
    "                        .filter(\"InChIKey IS NOT NULL\") \\\n",
    "                        .toPandas()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge Drug Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>structureChainId</th>\n",
       "      <th>queryGroupId</th>\n",
       "      <th>queryChainId</th>\n",
       "      <th>queryGroupNumber</th>\n",
       "      <th>queryAtomName</th>\n",
       "      <th>targetGroupId</th>\n",
       "      <th>targetChainId</th>\n",
       "      <th>targetGroupNumber</th>\n",
       "      <th>targetAtomName</th>\n",
       "      <th>distance</th>\n",
       "      <th>sequenceIndex</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ligand_id</th>\n",
       "      <th>formula_weight</th>\n",
       "      <th>inchi_key</th>\n",
       "      <th>DrugBankID</th>\n",
       "      <th>Name</th>\n",
       "      <th>CASNumber</th>\n",
       "      <th>DrugGroups</th>\n",
       "      <th>InChIKey</th>\n",
       "      <th>InChI</th>\n",
       "      <th>SMILES</th>\n",
       "      <th>Formula</th>\n",
       "      <th>KEGGCompoundID</th>\n",
       "      <th>KEGGDrugID</th>\n",
       "      <th>PubChemCompoundID</th>\n",
       "      <th>PubChemSubstanceID</th>\n",
       "      <th>ChEBIID</th>\n",
       "      <th>ChEMBLID</th>\n",
       "      <th>HETID</th>\n",
       "      <th>ChemSpiderID</th>\n",
       "      <th>BindingDBID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5EA2.E</td>\n",
       "      <td>FAD</td>\n",
       "      <td>G</td>\n",
       "      <td>301</td>\n",
       "      <td>C6</td>\n",
       "      <td>GLY</td>\n",
       "      <td>E</td>\n",
       "      <td>135</td>\n",
       "      <td>CA</td>\n",
       "      <td>15.728915</td>\n",
       "      <td>138</td>\n",
       "      <td>GPHMVGRRALIVLAHSERTSFNYAMKEAAAAALKKKGWEVVESDLY...</td>\n",
       "      <td>FAD</td>\n",
       "      <td>785.55</td>\n",
       "      <td>VWWQXMAJTJZDQX-UYBVJOGSSA-N</td>\n",
       "      <td>DB03147</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "      <td>146-14-5</td>\n",
       "      <td>approved</td>\n",
       "      <td>VWWQXMAJTJZDQX-UYBVJOGSSA-N</td>\n",
       "      <td>InChI=1S/C27H33N9O15P2/c1-10-3-12-13(4-11(10)2...</td>\n",
       "      <td>CC1=CC2=C(C=C1C)N(C[C@H](O)[C@H](O)[C@H](O)CO[...</td>\n",
       "      <td>C27H33N9O15P2</td>\n",
       "      <td>C00016</td>\n",
       "      <td>D00005</td>\n",
       "      <td>643975.0</td>\n",
       "      <td>46508543.0</td>\n",
       "      <td>16238.0</td>\n",
       "      <td>CHEMBL1232653</td>\n",
       "      <td>FAD</td>\n",
       "      <td>559059.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5EA2.G</td>\n",
       "      <td>FAD</td>\n",
       "      <td>E</td>\n",
       "      <td>301</td>\n",
       "      <td>O4B</td>\n",
       "      <td>PRO</td>\n",
       "      <td>G</td>\n",
       "      <td>68</td>\n",
       "      <td>CB</td>\n",
       "      <td>12.924247</td>\n",
       "      <td>71</td>\n",
       "      <td>GPHMVGRRALIVLAHSERTSFNYAMKEAAAAALKKKGWEVVESDLY...</td>\n",
       "      <td>FAD</td>\n",
       "      <td>785.55</td>\n",
       "      <td>VWWQXMAJTJZDQX-UYBVJOGSSA-N</td>\n",
       "      <td>DB03147</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "      <td>146-14-5</td>\n",
       "      <td>approved</td>\n",
       "      <td>VWWQXMAJTJZDQX-UYBVJOGSSA-N</td>\n",
       "      <td>InChI=1S/C27H33N9O15P2/c1-10-3-12-13(4-11(10)2...</td>\n",
       "      <td>CC1=CC2=C(C=C1C)N(C[C@H](O)[C@H](O)[C@H](O)CO[...</td>\n",
       "      <td>C27H33N9O15P2</td>\n",
       "      <td>C00016</td>\n",
       "      <td>D00005</td>\n",
       "      <td>643975.0</td>\n",
       "      <td>46508543.0</td>\n",
       "      <td>16238.0</td>\n",
       "      <td>CHEMBL1232653</td>\n",
       "      <td>FAD</td>\n",
       "      <td>559059.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  structureChainId queryGroupId queryChainId queryGroupNumber queryAtomName targetGroupId targetChainId  targetGroupNumber targetAtomName   distance  sequenceIndex                                           sequence ligand_id  formula_weight                    inchi_key DrugBankID                         Name CASNumber DrugGroups                     InChIKey                                              InChI                                             SMILES        Formula KEGGCompoundID KEGGDrugID  PubChemCompoundID  PubChemSubstanceID  ChEBIID       ChEMBLID HETID  ChemSpiderID  BindingDBID\n",
       "0           5EA2.E          FAD            G              301            C6           GLY             E                135             CA  15.728915            138  GPHMVGRRALIVLAHSERTSFNYAMKEAAAAALKKKGWEVVESDLY...       FAD          785.55  VWWQXMAJTJZDQX-UYBVJOGSSA-N    DB03147  Flavin adenine dinucleotide  146-14-5   approved  VWWQXMAJTJZDQX-UYBVJOGSSA-N  InChI=1S/C27H33N9O15P2/c1-10-3-12-13(4-11(10)2...  CC1=CC2=C(C=C1C)N(C[C@H](O)[C@H](O)[C@H](O)CO[...  C27H33N9O15P2         C00016     D00005           643975.0          46508543.0  16238.0  CHEMBL1232653   FAD      559059.0          NaN\n",
       "1           5EA2.G          FAD            E              301           O4B           PRO             G                 68             CB  12.924247             71  GPHMVGRRALIVLAHSERTSFNYAMKEAAAAALKKKGWEVVESDLY...       FAD          785.55  VWWQXMAJTJZDQX-UYBVJOGSSA-N    DB03147  Flavin adenine dinucleotide  146-14-5   approved  VWWQXMAJTJZDQX-UYBVJOGSSA-N  InChI=1S/C27H33N9O15P2/c1-10-3-12-13(4-11(10)2...  CC1=CC2=C(C=C1C)N(C[C@H](O)[C@H](O)[C@H](O)CO[...  C27H33N9O15P2         C00016     D00005           643975.0          46508543.0  16238.0  CHEMBL1232653   FAD      559059.0          NaN"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt = mt.merge(drugs, left_on=['inchi_key'], right_on=['InChIKey'], how='inner')\n",
    "#mt.to_csv(path_df+'drug_info.csv')\n",
    "mt.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_resName = []\n",
    "drug_dist = []\n",
    "drug_resNum = []\n",
    "drug_drugbankId = []\n",
    "drug_name =[]\n",
    "drug_inchi_key = []\n",
    "drug_CAS = []\n",
    "drug_ChEMBLID = []\n",
    "\n",
    "\n",
    "for i in df.index:\n",
    "    \n",
    "    p = df.loc[i,'structureChainId']\n",
    "    t = df.loc[i,'targetGroupNumber']\n",
    "    \n",
    "    r = ''\n",
    "    d = 1000\n",
    "    n = ''\n",
    "    m = ''\n",
    "    o = ''\n",
    "    q = ''\n",
    "    s = ''\n",
    "    u = ''\n",
    "    \n",
    "    tmp_drug_df = mt[(mt['structureChainId']==p)&(mt.targetGroupNumber==t)].sort_values('distance',ascending=True)\n",
    "    tmp_drug_df.reset_index(inplace=True)\n",
    "    \n",
    "    # TO DO: opt for additional ligands, not just nearest neighboring one\n",
    "    #if len(tmp_lig_df) > 0:\n",
    "    #    for j in tmp_lig_df.queryGroupId.unique():\n",
    "    #        tmp = tmp_lig_df[tmp_lig_df.queryGroupId==j].sort_values('distance',ascending=True)\n",
    "        \n",
    "    if len(tmp_drug_df) > 0:\n",
    "        \n",
    "        r = tmp_drug_df.loc[0,'queryGroupId']\n",
    "        d = tmp_drug_df.loc[0,'distance']\n",
    "        n = tmp_drug_df.loc[0,'queryGroupNumber']\n",
    "        m = tmp_drug_df.loc[0,'DrugBankID']\n",
    "        o = tmp_drug_df.loc[0,'Name']\n",
    "        q = tmp_drug_df.loc[0,'InChIKey']\n",
    "        s = tmp_drug_df.loc[0,'CASNumber']\n",
    "        u = tmp_drug_df.loc[0,'ChEMBLID']\n",
    "                   \n",
    "        try:\n",
    "            drug_resName.append(r)\n",
    "            drug_dist.append(d)\n",
    "            drug_resNum.append(n)\n",
    "            drug_drugbankId.append(m)\n",
    "            drug_name.append(o)\n",
    "            drug_inchi_key.append(q)\n",
    "            drug_CAS.append(s)\n",
    "            drug_ChEMBLID.append(u)\n",
    "            \n",
    "        except IndexError:\n",
    "            drug_resName.append('')\n",
    "            drug_dist.append(1000)\n",
    "            drug_resNum.append('')\n",
    "            drug_drugbankId.append('')\n",
    "            drug_name.append('')\n",
    "            drug_inchi_key.append('')\n",
    "            drug_CAS.append('')\n",
    "            drug_ChEMBLID.append('')\n",
    "            \n",
    "    else:\n",
    "        drug_resName.append('')\n",
    "        drug_dist.append(1000)\n",
    "        drug_resNum.append('')\n",
    "        drug_drugbankId.append('')\n",
    "        drug_name.append('')\n",
    "        drug_inchi_key.append('')\n",
    "        drug_CAS.append('')\n",
    "        drug_ChEMBLID.append('')\n",
    "        \n",
    "        \n",
    "df['Drug_dist'] = drug_dist\n",
    "df['Drug_resName'] = drug_resName\n",
    "df['Drug_resNum'] = drug_resNum\n",
    "df['Drug_DrugBankId'] = drug_drugbankId\n",
    "df['Drug_Name'] = drug_name\n",
    "df['Drug_Inchi'] = drug_inchi_key\n",
    "df['Drug_CAS'] = drug_CAS\n",
    "df['Drug_ChEMBLID'] = drug_ChEMBLID\n",
    "\n",
    "           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Protein_Change</th>\n",
       "      <th>Drug_dist</th>\n",
       "      <th>Drug_resName</th>\n",
       "      <th>Drug_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>VMCUB1_URINARY_TRACT</td>\n",
       "      <td>p.P28A</td>\n",
       "      <td>6.795431</td>\n",
       "      <td>FAD</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>KYAE1_OESOPHAGUS</td>\n",
       "      <td>p.V2V</td>\n",
       "      <td>7.047333</td>\n",
       "      <td>FAD</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>LS180_LARGE_INTESTINE</td>\n",
       "      <td>p.D143D</td>\n",
       "      <td>5.376522</td>\n",
       "      <td>FAD</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>MDAMB361_BREAST</td>\n",
       "      <td>p.E90Q</td>\n",
       "      <td>6.518649</td>\n",
       "      <td>FAD</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>697_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>p.G263D</td>\n",
       "      <td>7.827392</td>\n",
       "      <td>FAD</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>GRST_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>p.T335T</td>\n",
       "      <td>4.987658</td>\n",
       "      <td>FAD</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>MOLT16_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>p.T305T</td>\n",
       "      <td>3.126904</td>\n",
       "      <td>FAD</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>NCIH684_LIVER</td>\n",
       "      <td>p.I168T</td>\n",
       "      <td>3.739374</td>\n",
       "      <td>FAD</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>HUH1_LIVER</td>\n",
       "      <td>p.A122V</td>\n",
       "      <td>3.230645</td>\n",
       "      <td>FAD</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>JHUEM7_ENDOMETRIUM</td>\n",
       "      <td>p.R46Q</td>\n",
       "      <td>8.949179</td>\n",
       "      <td>FAD</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>DND41_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>p.L251L</td>\n",
       "      <td>8.917463</td>\n",
       "      <td>FAD</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>JHUEM7_ENDOMETRIUM</td>\n",
       "      <td>p.L145I</td>\n",
       "      <td>7.641257</td>\n",
       "      <td>FAD</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Hugo_Symbol                       Tumor_Sample_Barcode Protein_Change  Drug_dist Drug_resName                    Drug_Name\n",
       "11      TXNRD1                       VMCUB1_URINARY_TRACT         p.P28A   6.795431          FAD  Flavin adenine dinucleotide\n",
       "13      TXNRD1                           KYAE1_OESOPHAGUS          p.V2V   7.047333          FAD  Flavin adenine dinucleotide\n",
       "15      TXNRD1                      LS180_LARGE_INTESTINE        p.D143D   5.376522          FAD  Flavin adenine dinucleotide\n",
       "20      TXNRD1                            MDAMB361_BREAST         p.E90Q   6.518649          FAD  Flavin adenine dinucleotide\n",
       "24      TXNRD1     697_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE        p.G263D   7.827392          FAD  Flavin adenine dinucleotide\n",
       "27      TXNRD1    GRST_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE        p.T335T   4.987658          FAD  Flavin adenine dinucleotide\n",
       "28      TXNRD1  MOLT16_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE        p.T305T   3.126904          FAD  Flavin adenine dinucleotide\n",
       "31      TXNRD1                              NCIH684_LIVER        p.I168T   3.739374          FAD  Flavin adenine dinucleotide\n",
       "34      TXNRD1                                 HUH1_LIVER        p.A122V   3.230645          FAD  Flavin adenine dinucleotide\n",
       "36      TXNRD1                         JHUEM7_ENDOMETRIUM         p.R46Q   8.949179          FAD  Flavin adenine dinucleotide\n",
       "42      TXNRD1   DND41_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE        p.L251L   8.917463          FAD  Flavin adenine dinucleotide\n",
       "43      TXNRD1                         JHUEM7_ENDOMETRIUM        p.L145I   7.641257          FAD  Flavin adenine dinucleotide"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.Drug_dist<10][['Hugo_Symbol','Tumor_Sample_Barcode','Protein_Change','Drug_dist','Drug_resName','Drug_Name']].drop_duplicates().head(12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save Final DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(output_file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Shutdown Spark\n",
    "spark.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Protein_Change</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>variationId</th>\n",
       "      <th>uniprotId</th>\n",
       "      <th>structureChainId</th>\n",
       "      <th>targetGroupId</th>\n",
       "      <th>targetChainId</th>\n",
       "      <th>targetGroupNumber</th>\n",
       "      <th>cluster_x</th>\n",
       "      <th>cluster_y</th>\n",
       "      <th>cluster_z</th>\n",
       "      <th>atom_nom</th>\n",
       "      <th>PDB_CHAIN_ID</th>\n",
       "      <th>SP_PRIMARY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>coverage</th>\n",
       "      <th>PTM_dist</th>\n",
       "      <th>PTM_type</th>\n",
       "      <th>PTM_resName</th>\n",
       "      <th>PTM_resNum</th>\n",
       "      <th>Ligand_dist</th>\n",
       "      <th>Ligand_resName</th>\n",
       "      <th>Ligand_resNum</th>\n",
       "      <th>Drug_dist</th>\n",
       "      <th>Drug_resName</th>\n",
       "      <th>Drug_resNum</th>\n",
       "      <th>Drug_DrugBankId</th>\n",
       "      <th>Drug_Name</th>\n",
       "      <th>Drug_Inchi</th>\n",
       "      <th>Drug_CAS</th>\n",
       "      <th>Drug_ChEMBLID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>3758</td>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>p.A122V</td>\n",
       "      <td>HUH1_LIVER</td>\n",
       "      <td>chr12:g.104713303C&gt;T</td>\n",
       "      <td>Q16881</td>\n",
       "      <td>3QFB.A</td>\n",
       "      <td>ALA</td>\n",
       "      <td>A</td>\n",
       "      <td>160</td>\n",
       "      <td>58.463001</td>\n",
       "      <td>-93.374001</td>\n",
       "      <td>43.442001</td>\n",
       "      <td>CB</td>\n",
       "      <td>3QFB.A</td>\n",
       "      <td>Q16881</td>\n",
       "      <td>TXNRD1</td>\n",
       "      <td>0.762712</td>\n",
       "      <td>8.518215</td>\n",
       "      <td>Ubiquitination</td>\n",
       "      <td>PHE</td>\n",
       "      <td>157</td>\n",
       "      <td>3.230645</td>\n",
       "      <td>FAD</td>\n",
       "      <td>600</td>\n",
       "      <td>3.230645</td>\n",
       "      <td>FAD</td>\n",
       "      <td>600</td>\n",
       "      <td>DB03147</td>\n",
       "      <td>Flavin adenine dinucleotide</td>\n",
       "      <td>VWWQXMAJTJZDQX-UYBVJOGSSA-N</td>\n",
       "      <td>146-14-5</td>\n",
       "      <td>CHEMBL1232653</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index Hugo_Symbol Protein_Change Tumor_Sample_Barcode           variationId uniprotId structureChainId targetGroupId targetChainId  targetGroupNumber  cluster_x  cluster_y  cluster_z atom_nom PDB_CHAIN_ID SP_PRIMARY    GENE  coverage  PTM_dist        PTM_type PTM_resName PTM_resNum  Ligand_dist Ligand_resName Ligand_resNum  Drug_dist Drug_resName Drug_resNum Drug_DrugBankId                    Drug_Name                   Drug_Inchi  Drug_CAS  Drug_ChEMBLID\n",
       "34   3758      TXNRD1        p.A122V           HUH1_LIVER  chr12:g.104713303C>T    Q16881           3QFB.A           ALA             A                160  58.463001 -93.374001  43.442001       CB       3QFB.A     Q16881  TXNRD1  0.762712  8.518215  Ubiquitination         PHE        157     3.230645            FAD           600   3.230645          FAD         600         DB03147  Flavin adenine dinucleotide  VWWQXMAJTJZDQX-UYBVJOGSSA-N  146-14-5  CHEMBL1232653"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene = 'TXNRD1'\n",
    "struct = '3QFB.A'\n",
    "df[(df.Hugo_Symbol==gene)&(df.structureChainId==struct)&(df.Protein_Change=='p.A122V')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
